internal/logger/logrotate.go

```diff
@@ -86,40 +92,106 @@ func (w *Writer) Close() error {
 	return nil
 }
 
+var stdErrEnc = json.NewEncoder(os.Stderr)
+
 func (w *Writer) listen() {
 	for {
 		var r io.Reader = w.pr
 		if w.opts.MaximumFileSize > 0 {
 			r = io.LimitReader(w.pr, w.opts.MaximumFileSize)
 		}
 		if _, err := io.Copy(w.f, r); err != nil {
-			fmt.Println("Failed to write to log file", err)
+			msg := fmt.Sprintf("unable to write to log file %v: %w", w.f.Name(), err)
+			stdErrEnc.Encode(&log.Entry{
+				Level:   ErrorKind,
+				Message: msg,
+				Time:    time.Now().UTC(),
+				Trace:   &log.Trace{Message: msg},
+			})
 		}
 		if err := w.rotate(); err != nil {
-			fmt.Println("Failed to rotate log file", err)
+			msg := fmt.Sprintf("unable to rotate log file %v: %w", w.f.Name(), err)
+			stdErrEnc.Encode(&log.Entry{
+				Level:   ErrorKind,
+				Message: msg,
+				Time:    time.Now().UTC(),
+				Trace:   &log.Trace{Message: msg},
+			})
 		}
 	}
 }
 
 func (w *Writer) closeCurrentFile() error {
 	if err := w.f.Close(); err != nil {
-		return fmt.Errorf("failed to close current log file: %w", err)
+		return fmt.Errorf("unable to close current log file: %w", err)
 	}
 
 	return nil
 }
 
+func (w *Writer) compress() error {
+	if !w.opts.Compress {
+		return nil
+	}
+
+	oldLgFile := w.f.Name()
+	r, err := os.Open(oldLgFile)
+	if err != nil {
+		return err
+	}
+
+	gw, err := os.Create(oldLgFile + ".gz")
+	if err != nil {
+		return err
+	}
+
+	var wc io.WriteCloser
+	wc = gzip.NewWriter(gw)
+	if _, err = io.Copy(wc, r); err != nil {
+		return err
+	}
+
+	// close everything before we delete.
```

Comment:
There are couple of things missing here.
1. We should call `gzip.Writer`'s Close method before we call fsync on the `.gz` file.
2. `gzip.Writer.Close` method doesn't call `Close` on the underlying `io.WriteCloser`, it simply flushes any pending compressed data. We need to call `gw.Close` explicitly.